docs: add a howto for configuring UFW (#1642)
canonical/k8s-snap: docs/canonicalk8s/snap/howto/networking/ufw.md

# How to configure Uncomplicated Firewall (UFW)

In this howto we present a set of firewall rules/guidelines you should
consider when setting up {{product}}. Be aware that these rules
may be incompatible with your network setup, you may find them too permissive
or too restrictive. We recommend you review them and tune them to match
your needs. Also, be aware that the firewall rules need to be reviewed

> for each service hosted in Kubernetes as there might be special requirements.

Author: Done

## Prerequisites

This guide assumes the following:

- A machine with Ubuntu.

> - You have root or sudo access to the machine.

Author: Good point. Have a look at the updated version of this section.

## Install and enable UFW

> To install Uncomplicated Firewall:

Author: Yes, UFW needs to be installed on all nodes.

 
Allow traffic on port 6443 with:

``` sh
sudo ufw allow 6443/tcp
```

Services such as for example the metrics-server need access to the kubelet,
controller manager and kube scheduler to query for metrics.
Kubelet runs on all nodes, while the kube-controller-manager and
kube-scheduler run only on the control plane nodes:

Allow traffic on port 10250 on all nodes:

```sh
sudo ufw allow 10250/tcp
```

Allow traffic on ports 10257 and 10259 on control plane nodes:

```sh
sudo ufw allow 10257/tcp
sudo ufw allow 10259/tcp
```

## Allow cluster formation

To form an HA cluster the datastore used by Kubernetes (dqlite/etcd) needs
to establish a direct connection among its peers. In dqlite this is done
through port 9000 while on etcd port 2380 is used.

Allow traffic on port 9000 on control plane nodes with dqlite:

```sh
sudo ufw allow 9000/tcp
```

Allow traffic on port 2380 on control plane nodes with etcd:

```sh
sudo ufw allow 2380/tcp
```

Cluster formation is overseen by a Kubernetes daemon running on all nodes
on port 6400.

Allow traffic on port 6400:

```sh
sudo ufw allow 6400/tcp
```

## Allow CNI specific communication

The default CNI used in {{product}} is cilium.

> Unless you are not disabling this network plugin and deploying your own,

Author: Thank you.

you should consider the following firewall rules.

Allow cluster-wide Cilium agent health checks and VXLAN traffic:

```sh
sudo ufw allow 4240/tcp
sudo ufw allow 8472/udp
```

## UFW troubleshooting

The [ports-and-services] page has a list of all ports {{product}} uses.

To inspect a failing service you can enable logging:

```sh
sudo ufw logging on
```

Monitor the firewall logs with:

```sh
tail -f /var/log/ufw.log
```

> At the end disable logging:

Author: Yes one, could argue that the resources used are high.

# How to configure Uncomplicated Firewall (UFW)

> In this how-to we present a set of firewall rules/guidelines
> you should consider when setting up {{product}}.

Reviewer: ```suggestion
This how-to presents a set of firewall rules/guidelines
that should be considered when setting up {{product}}.
```
(let's use an active voice)

Be aware that these rules may be incompatible with your network setup

> and we recommend you review and tune them to match your needs.  

Reviewer: This sounds like a note can we format it accordingly please?
So starting with: ```{note}

Author: This is a good point but I did not like how the how-to would start with the note banner so I rephrased this part a bit so it does not sound like a warning.

> Be aware that these rules may be incompatible with your network setup
> and we recommend you review and tune them to match your needs.  

Reviewer: ```suggestion
Please be aware that these rules may be incompatible with your network setup, so we recommend reviewing and adjusting them to match your needs.  
```
(active voice + causal conjunction)

> Also, be aware that for each service hosted in Kubernetes,
> the firewall rules need to be reviewed as there might be
> special requirements for each.

Reviewer: ```suggestion
Also, please review the firewall rules for each service hosted in Kubernetes, as there might be
special requirements for each.
```

## Prerequisites

This guide assumes the following:

> - A machine with Ubuntu where you have installed
>   or you plan to install {{product}}.
> - You have root or sudo access to the machine.

Reviewer: ```suggestion
- An ubuntu machine where {{product}} is installed or will be installed.
- Root or sudo access to the machine.
```
(active voice)

## Install and enable UFW 

Uncomplicated Firewall needs to be configured on all nodes of {{product}}.

> To do so try:

Reviewer: ```suggestion
To do this, run:
```

```sh
sudo apt update
sudo apt install ufw
```

To verify UFW is installed try:

```sh
sudo ufw status verbose
```

> If you need to maintain ssh access to the machine, make sure you configure
> UFW to allow `OpenSSH` before enabling it:

Reviewer: ```suggestion
To maintain SSH access to the machine, allow `OpenSSH` through UFW before enabling the firewall:
```

```sh
sudo ufw allow OpenSSH
```

Now you are ready to enable UFW:

```sh
sudo ufw enable
```

## Allow package forwarding

> Forwarding is needed because containers typically live in isolated networks

Reviewer: ```suggestion
Package forwarding is needed because containers typically live in isolated networks
```

and expect the host-to-route traffic between their internal network and the
outside world to be allowed.

> First edit `/etc/default/ufw` and allow UFW to route/forward packets:

Reviewer: ```suggestion
First, edit `/etc/default/ufw` and allow UFW to route/forward packets:
```

```sh
DEFAULT_FORWARD_POLICY="ACCEPT"
```

> 
> Enable IP forwarding by editing `/etc/sysctl.conf` so it persists through
> system reboots:
> 
> ```sh
> net.ipv4.ip_forward=1
> ```
> 
> Or use `sysctl` directly so forwarding is applied immediately,
> without rebooting the system:
> 

Reviewer: ```suggestion

If you want the forwarding rules to persist through system reboots, enable IP forwarding by editing `/etc/sysctl.conf`:

```sh
net.ipv4.ip_forward=1
```

Otherwise, use `sysctl` directly to apply the forwarding rules immediately
without rebooting the system:

```
(Let's make the two options more clearer in the wording)

```sh
sudo sysctl -w net.ipv4.ip_forward=1
```

> Reload UFW:

Reviewer: ```suggestion
Reload UFW to apply the changes:
```

```sh
sudo ufw reload
```

> 

Reviewer: We could optionally add a verification step here using ufw status. Wdyt?

## Allow access to the Kubernetes services

> Services such as CoreDNS require access to the Kubernetes API

Reviewer: We could add a link to CoreDNS docs here.

Author: But do we want readers to click on CoreDNS from here? There is nothing for them to do on CoreDNS, it is just an example of a service.

Reviewer: That's fair, they don't need extra information about CoreDNS for this guide.

server listening on port 6443.
 
Allow traffic on port 6443 with:

``` sh
sudo ufw allow 6443/tcp
```

Services such as the metrics-server need access to the kubelet,
controller manager and kube scheduler to query for metrics.

Kubelet runs on all nodes, so allow traffic on port 10250 on all nodes:

```sh
sudo ufw allow 10250/tcp
```

> The kube-controller-manager and kube-scheduler run only on
> the control plane so allow traffic on ports 10257 and 10259
> on control plane nodes:

Reviewer: ```suggestion
The kube-controller-manager and kube-scheduler only run on
the control plane, therefore permit traffic on ports 10257 and 10259
of the control plane nodes:
```

```sh
sudo ufw allow 10257/tcp
sudo ufw allow 10259/tcp
```

## Allow cluster formation

To form a High Availability (HA) cluster the datastore used by Kubernetes

> (dqlite/etcd) needs to establish a direct connection among its peers.

Reviewer: @nhennigan  We capitalize Dqlite in the rest of the docs but keep etcd lower-case. I think we should be uniform in out capitalization with etcd/dqlite. Wdyt?

Reviewer 2: Good catch. Yes Dqlite should be capitalized according to their docs. I think the confusion stems from the fact on github it is listed with a lower case. etcd is the name of the product and they do not capitalize

> In dqlite this is done through port 9000 while on etcd port 2380 is used.

Author: Done

Allow traffic on port 9000 on control plane nodes with dqlite:

```sh
sudo ufw allow 9000/tcp
```

> Allow traffic on port 2380 on control plane nodes with etcd:

Reviewer: We are actually not using upstreams port but rather by default:
```
EtcdPort:      utils.Pointer(12379),
EtcdPeerPort:  utils.Pointer(12380),
```

```sh
sudo ufw allow 2380/tcp

> ```

Reviewer: What about port 2379 for client communication?

Cluster formation is overseen by a Kubernetes daemon running on all nodes
on port 6400.

> Allow traffic on port 6400:

Reviewer: ```suggestion
Open port 6400 to permit cluster formation traffic:
```

```sh
sudo ufw allow 6400/tcp
```

> ## Allow CNI specific communication

Reviewer: ```suggestion
## Enable CNI specific communication
```
(we've repeated allow a lot)

> If you are using the default network plugin (Cilium),
> you should consider the following firewall rules.

Reviewer: ```suggestion
When using the default network plugin (Cilium),
 consider the following firewall rules.
```

Allow cluster-wide Cilium agent health checks and VXLAN traffic:

```sh
sudo ufw allow 4240/tcp
sudo ufw allow 8472/udp
```

## UFW troubleshooting

The [ports-and-services] page has a list of all ports {{product}} uses.

To inspect a failing service you can enable logging:

```sh
sudo ufw logging on
```

Monitor the firewall logs with:

```sh
tail -f /var/log/ufw.log
```

The logs will show you which packets are dropped, their destination
and source as well as the protocol used and the destination port.

> This information may help you identify any other ports or services

Reviewer: ```suggestion
This information helps to identify any other ports or services
```

you need to enable within UFW.

> To keep the resources used by UFW to a minimum you can disable logging:

Reviewer: ```suggestion
After troubleshooting, keep the resources used by UFW to a minimum by disabling logging again:
```

# How to configure Uncomplicated Firewall (UFW)

This how-to presents a set of firewall rules/guidelines
that should be considered when setting up {{product}}.
These rules may be incompatible with your network setup,
so we recommend you review and tune them to match your needs.

## Prerequisites

This guide assumes the following:

> - An ubuntu machine where {{product}} is installed or will be installed.

Reviewer 2: ```suggestion
- An Ubuntu machine where {{product}} is installed or will be installed.
```
